scripts/check_ecosystem.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+"""Check two versions of ruff against a corpus of open-source code.
+
+Example usage:
+
+    scripts/check_ecosystem.py <path/to/ruff1> <path/to/ruff2>
+"""
+
+# ruff: noqa: T201
+
+import argparse
+import asyncio
+import difflib
+import heapq
+import re
+import tempfile
+from asyncio.subprocess import PIPE, create_subprocess_exec
+from contextlib import asynccontextmanager
+from pathlib import Path
+from typing import TYPE_CHECKING, NamedTuple, Self
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncIterator, Iterator, Sequence
+
+
+class Repository(NamedTuple):
+    """A GitHub repository at a specific ref."""
+
+    org: str
+    repo: str
+    ref: str
+
+    @asynccontextmanager
+    async def clone(self: Self) -> "AsyncIterator[Path]":
+        """Shallow clone this repository to a temporary directory."""
+        with tempfile.TemporaryDirectory() as tmpdir:
+            process = await create_subprocess_exec(
+                "git",
+                "clone",
+                "--config",
+                "advice.detachedHead=false",
+                "--quiet",
+                "--depth",
+                "1",
+                "--no-tags",
+                "--branch",
+                self.ref,
+                f"https://github.com/{self.org}/{self.repo}",
+                tmpdir,
+            )
+
+            await process.wait()
+
+            yield Path(tmpdir)
+
+
+REPOSITORIES = {
+    "zulip": Repository("zulip", "zulip", "main"),
+    "bokeh": Repository("bokeh", "bokeh", "branch-3.2"),
```

Comment:
I'm happy to amend this list after merging, but others that come to mind:

- https://github.com/scikit-build/scikit-build/blob/main/pyproject.toml
- https://github.com/apache/airflow/blob/main/pyproject.toml

Not sure what the incremental cost is for each additional repo. I guess the two costs are "job time" and "job output" -- e.g., if we make a mistake, we might see it repeated across repos. But here, I'm thinking back to projects that we've broken in the past, and that have reasonably advanced configurations.

Reply:
I think this would be better as a data file (yaml or toml or something), it would make it easier to maintain. The cost per repo would be very interesting, I'd be happy to add a few of the projects I work on (and also am aware of). :P Pretty sure it's low because Ruff is fast? (Usually this stops me from setting up something like this for things like pybind11 and scikit-build - though scikit-build is planned anyway).

I'd also avoid showing any repos that aren't broken in the default output. I think with that you could scale up to at least a few dozen. Mypy primer does something ike 50. (Also, it hardcodes the values in just like this, and doesn't use a config file for it).